README.md
# pull_to_refresh

A widget that provided to  pull-up load and pull-down refresh with Flutter.

## Features

* pull up and pull down
* Simple interface, easy access
* support most of the component in Flutter such as scroll and non-scroll component


## Show cases

<img src="https://github.com/bytedance/pull_to_refresh/blob/master/doc/image.gif" width="320">


## Quick Start
1. Add ```import package:pull_to_refresh/pull_to_refresh_widget.dart; ```
2. Using PullToRefreshWidget wrap outside your content widget
```
PullToRefreshWidget(
    isRefreshEnable: true,
    headerBuilder: _buildHeaderWidget,
    onRefresh: _handleRefresh,
    isLoadMoreEnable: hasMore,
    onLoadMore: _onLoadMore,
    footerBuilder: _buildFootWidget,
    child: ListView.builder(
        itemCount: items.length,
        itemBuilder: (context, index) {
        return new ListTile(
            title: Text("Index$index"),
        );
    }));
```

3. You should set the headerBuilder and footerBuilder in PullToRefreshWidget that generate header widget and footer widget. 
The header widget must inherit DefaultRefreshHeaderWidget.
```
DefaultRefreshHeaderWidget _buildHeaderWidget(BuildContext context) {
    return DefaultRefreshHeaderWidget();
  }

Widget _buildFootWidget(BuildContext context) {
    return new Container(
      alignment: Alignment.center,
      height: 60.0,
      child: new Center(
        child: new Row(
            textDirection: TextDirection.ltr,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              new Text("正在加载中。。。"),
              new CircularProgressIndicator(strokeWidth: 2.0)
            ]),
      ),
    );
  }
```
4. The onRefresh and onLoadMore will be callback when the indicator state is refreshed and load more. 
The function returned [Future] must complete when the refresh operation is finished.

```
  Future<Null> _handleRefresh() async {
    await Future.delayed(Duration(seconds: 50), () {
      setState(() {
        items.clear();
        items = List.generate(40, (i) => i);
      });
    });
  }

  Future<Null> _onLoadMore() async {
    await Future.delayed(Duration(seconds: 50), () {
      setState(() {
        if (index == 2) {
          hasMore = false;
        }
        int length = items.length;
        index++;
        items.addAll(List.generate(16, (i) => i + length));
      });
    });
  }
```

full example see here: main.dart.




CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* 初始化工程



lib/proxy_header_widget.dart
import 'package:flutter/material.dart';
import 'package:pull_to_refresh/refresh_header_widget.dart';
import 'package:pull_to_refresh/pull_to_refresh_widget.dart';

const Duration _kSizeDuration = Duration(milliseconds: 200);

RefreshObserver of(BuildContext context) {
  /// 这个方法内，调用 context.inheritFromWidgetOfExactType
  InheritedWidget widget =
      context.inheritFromWidgetOfExactType(_InheritedRefreshContainer);
  if (widget is _InheritedRefreshContainer) {
    return widget.observer;
  } else {
    return null;
  }
}

/// 刷新头的代理类
class ProxyIndicatorWidget extends StatefulWidget {
  final HeaderBuilder builder;
  final double height;

  const ProxyIndicatorWidget({Key key, this.builder, this.height})
      : super(key: key);

  @override
  State createState() {
    return new _ProxyIndicatorState();
  }
}

class _ProxyIndicatorState extends State<ProxyIndicatorWidget>
    with TickerProviderStateMixin, RefreshListener {
  AnimationController _sizeController;
  RefreshObserver observer;

  @override
  void initState() {
    _sizeController = new AnimationController(
        vsync: this, lowerBound: 0.000001, duration: _kSizeDuration);
    observer = new RefreshObserver();
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
    _sizeController.dispose();
    observer = null;
    super.initState();
  }

  @override
  void reset() {
    if (observer == null ||
        observer.callbacks == null ||
        observer.callbacks.isEmpty) {
      return;
    }
    for (RefreshListener listener in observer.callbacks) {
      listener.reset();
    }

    _sizeController.animateTo(0.0);
  }

  @override
  void refreshing() {
    _sizeController.value = 1.0;
    if (observer == null ||
        observer.callbacks == null ||
        observer.callbacks.isEmpty) {
      return;
    }
    for (RefreshListener listener in observer.callbacks) {
      listener.refreshing();
    }
  }

  @override
  Widget build(BuildContext context) {
    return new Column(
      children: <Widget>[
        new SizeTransition(
          sizeFactor: _sizeController,
          child: new Container(height: widget.height),
        ),
        _InheritedRefreshContainer(
            observer: observer, child: widget.builder(context)),
      ],
    );
  }

  @override
  void onMoving(double offset, double totalHeight, double maxDragHeight) {
    if (observer == null ||
        observer.callbacks == null ||
        observer.callbacks.isEmpty) {
      return;
    }
    for (RefreshListener listener in observer.callbacks) {
      listener.onMoving(offset, totalHeight, maxDragHeight);
    }
  }
}

/// 刷新接口的观察者
class RefreshObserver {
  /// 注册被观察者的列表
  List<RefreshListener> callbacks = List();

  addListener(RefreshListener listener) {
    if (listener != null && !callbacks.contains(listener)) {
      callbacks.add(listener);
    }
  }

  removeListener(RefreshListener listener) {
    callbacks.remove(listener);
  }
}

class _InheritedRefreshContainer extends InheritedWidget {
  final RefreshObserver observer;

  /// 我们知道InheritedWidget总是包裹的一层，所以它必有child
  _InheritedRefreshContainer(
      {Key key, @required this.observer, @required Widget child})
      : super(key: key, child: child);

  /// 参考MediaQuery,这个方法通常都是这样实现的。如果新的值和旧的值不相等，就需要notify
  @override
  bool updateShouldNotify(_InheritedRefreshContainer oldWidget) =>
      observer != oldWidget.observer;
}


lib/refresh_header_widget.dart
import 'package:flutter/material.dart';
import 'package:pull_to_refresh/proxy_header_widget.dart';

/// 下拉刷新头部 widget
abstract class RefreshHeaderWidget<T extends RefreshState>
    extends StatefulWidget {
  const RefreshHeaderWidget({Key key}) : super(key: key);

  @override
  RefreshState createState();
}

abstract class RefreshState<T extends StatefulWidget> extends State<T>
    implements RefreshListener {
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    RefreshObserver observer = of(context);

    /// 添加监听事件
    if (observer != null) {
      observer.addListener(this);
    }
  }

  @override
  void dispose() {
    RefreshObserver observer = of(context);

    /// 移除监听事件
    if (observer != null) {
      observer.removeListener(this);
    }
    super.dispose();
  }
}

/// 监听下拉刷新状态接口
abstract class RefreshListener {
  /// 开始刷新，Header 进入刷新状态之前调用。
  void refreshing();

  /// 手指拖动下拉
  /// [offset] 下拉的像素偏移量
  /// height 高度 HeaderHeight or FooterHeight
  /// maxDragHeight 最大拖动高度
  void onMoving(double offset, double totalHeight, double maxDragHeight);

  /// 重置，回到初始状态
  void reset();
}


lib/pull_to_refresh_widget.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:pull_to_refresh/refresh_header_widget.dart';
import 'package:pull_to_refresh/proxy_header_widget.dart';
import 'dart:math' as math;

/// 刷新状态
enum _RefreshStatus {
  /// 初始化状态
  reset,

  /// 下拉可以刷新
  pullToRefresh,

  /// 松开后刷新
  releaseToRefresh,

  /// 刷新中，此时手指已离开屏幕
  refreshing,
}

/// 创建下拉头部视图函数
typedef HeaderBuilder = RefreshHeaderWidget Function(BuildContext context);

/// 创建上拉底部视图
typedef FooterBuilder = Widget Function(BuildContext context);

/// 下拉刷新回调
typedef Future<void> OnRefreshCallback();

/// 加载更多回调
typedef Future<void> OnLoadMoreCallback();

class PullToRefreshWidget extends StatefulWidget {
  /// 下拉头部视图创建者
  final HeaderBuilder headerBuilder;

  /// 上拉底部视图创建者
  final FooterBuilder footerBuilder;

  /// 是否开启上拉加载更多
  final bool isLoadMoreEnable;

  /// 是否下拉刷新
  final bool isRefreshEnable;

  /// 监听下拉刷新属性
  final OnRefreshCallback onRefresh;

  /// 监听上拉加载更多
  final OnLoadMoreCallback onLoadMore;

  /// 内容视图
  final Widget child;

  /// 触发刷新距离 与 HeaderHeight 的比率
  final double headerTriggerRate;

  const PullToRefreshWidget(
      {Key key,
      @required this.child,
      this.headerTriggerRate = 1.0,
      this.onRefresh,
      this.onLoadMore,
      this.headerBuilder,
      this.isRefreshEnable = false,
      this.footerBuilder,
      this.isLoadMoreEnable = false})
      : assert(child != null),
        super(key: key);

  @override
  State createState() {
    return new _PullRefreshState();
  }
}

class _PullRefreshState extends State<PullToRefreshWidget> {
  /// 下拉刷新头部控件的高度
  double _headerHeight = 0.0;

  /// 触发刷新的最大距离
  double _maxDistance = 0.0;

  /// 头部控件的key,用来获取头部的高度
  final GlobalKey _headerKey = new GlobalKey();

  /// 下拉刷新的刷新状态
  _RefreshStatus _headerState = _RefreshStatus.reset;

  /// 是否正在加载更多，防止来回滑动导致接口调用多次的问题
  bool _isLoadingMore;

  /// scrollview的滑动监听者
  ScrollController _scrollController;

  @override
  void initState() {
    _scrollController = new ScrollController();
    _scrollController.addListener(() {
      if (!widget.isLoadMoreEnable) {
        return;
      }

      ///当滑动的距离大于可滑动的最大像素时回调下拉刷新接口
      if (_scrollController.position.pixels >=
          _scrollController.position.maxScrollExtent) {
        /// 防止回调多次
        if (_isLoadingMore == true) {
          return;
        }
        if (widget.onLoadMore != null) {
          _isLoadingMore = true;
          widget.onLoadMore().whenComplete(() {
            _isLoadingMore = false;
          });
        }
      }
    });

    WidgetsBinding.instance.addPostFrameCallback((duration) {
      setState(() {
        if (widget.isRefreshEnable) {
          _headerHeight = _headerKey.currentContext.size.height;
          _maxDistance = widget.headerTriggerRate * _headerHeight;
        }
      });
    });
    super.initState();
  }

  @override
  void dispose() {
    if (_scrollController != null) {
      _scrollController.dispose();
    }
    super.dispose();
  }

  /// 下拉刷新头是否合法
  bool _isHeaderValid() {
    return widget.isRefreshEnable &&
        (_headerKey.currentState is RefreshListener);
  }

  bool _onNotification(ScrollNotification notification) {
    if (!_isHeaderValid()) {
      return false;
    }

    /// 滚动区域超过屏幕时，不进行处理
    if (!(notification.metrics.pixels < 0) &&
        !(notification.metrics.pixels > 0)) {
      return false;
    }

    /// 正在滑动中，此时滑动位置会变化
    if (notification is ScrollUpdateNotification) {
      /// notification.dragDetails 表明此时我们的手指已经离开了屏幕
      if (notification.dragDetails == null) {
        _handleScrollEnd(notification);
      } else {
        _handleScrollUpdate(notification);
      }
    } else if (notification is ScrollEndNotification) {
      /// 滑动结束
      _handleScrollEnd(notification);
    }
    return false;
  }

  /// 处理正在滑动中
  bool _handleScrollUpdate(ScrollUpdateNotification notification) {
    if (notification.metrics.pixels <= notification.metrics.minScrollExtent) {
      if (!widget.isRefreshEnable) {
        return false;
      }
    } else {
      return false;
    }

    /// 已经处于正在刷新状态，就直接返回
    if (_headerState == _RefreshStatus.refreshing) {
      return false;
    }

    double offset =
        notification.metrics.minScrollExtent - notification.metrics.pixels;
    RefreshListener refreshListener =
        _headerKey.currentState as RefreshListener;

    refreshListener.onMoving(offset, _headerHeight, _maxDistance);

    if (_headerState == _RefreshStatus.pullToRefresh &&
        offset >= _maxDistance) {
      _headerState = _RefreshStatus.releaseToRefresh;
    } else if (_headerState != _RefreshStatus.pullToRefresh &&
        offset < _maxDistance) {
      _headerState = _RefreshStatus.pullToRefresh;
    }
    return false;
  }

  bool _handleScrollEnd(ScrollNotification notification) {
    if (!_isHeaderValid()) {
      return false;
    }
    if (!(_headerKey.currentState is RefreshListener)) {
      return false;
    }

    RefreshListener callback = _headerKey.currentState as RefreshListener;

    /// 松开后刷新
    if (_headerState == _RefreshStatus.releaseToRefresh &&
        widget.onRefresh != null) {
      _headerState = _RefreshStatus.refreshing;
      callback.refreshing();

      final Future<void> refreshResult = widget.onRefresh();
      if (refreshResult == null) {
        return false;
      }
      refreshResult.whenComplete(() {
        _headerState = _RefreshStatus.reset;
        callback.reset();
      });

      return false;
    }

    if (_headerState == _RefreshStatus.refreshing) return false;
    callback.reset();
    return false;
  }

  @override
  Widget build(BuildContext context) {
    List<Widget> children = new List();
    if (widget.isRefreshEnable) {
      Widget header = widget.headerBuilder != null
          ? new ProxyIndicatorWidget(
              builder: widget.headerBuilder,
              key: _headerKey,
              height: _headerHeight,
            )
          : new Container();
      children.add(SliverToBoxAdapter(
        child: header,
      ));
    }

    if (widget.child is ScrollView) {
      children.addAll((widget.child as ScrollView).buildSlivers(context));
    } else {
      children.add(SliverToBoxAdapter(
        child: widget.child,
      ));
    }

    if (widget.isLoadMoreEnable && widget.footerBuilder != null) {
      children.add(SliverToBoxAdapter(child: widget.footerBuilder(context)));
    }

    return new LayoutBuilder(builder: (context, cons) {
      return new Stack(
        children: <Widget>[
          new Positioned(
              top: -_headerHeight,
              bottom: 0,
              left: 0.0,
              right: 0.0,
              child: new NotificationListener(
                child: new CustomScrollView(
                  controller: _scrollController,
                  physics: widget.isRefreshEnable
                      ? new _PullToRefreshScrollPhysics()
                      : null,
                  slivers: children,
                ),
                onNotification: _onNotification,
              )),
        ],
      );
    });
  }
}

///copy from BouncingScrollPhysics ,用来控制下拉刷新的滑动特性
class _PullToRefreshScrollPhysics extends ScrollPhysics {
  const _PullToRefreshScrollPhysics({ScrollPhysics parent})
      : super(parent: parent);

  @override
  _PullToRefreshScrollPhysics applyTo(ScrollPhysics ancestor) {
    return _PullToRefreshScrollPhysics(parent: buildParent(ancestor));
  }

  double frictionFactor(double overscrollFraction) =>
      0.52 * math.pow(1 - overscrollFraction, 2);

  @override
  bool shouldAcceptUserOffset(ScrollMetrics position) {
    // TODO: implement shouldAcceptUserOffset
    return true;
  }

  @override
  double applyPhysicsToUserOffset(ScrollMetrics position, double offset) {
    assert(offset != 0.0);
    assert(position.minScrollExtent <= position.maxScrollExtent);

    if (!position.outOfRange) return offset;

    final double overscrollPastStart =
        math.max(position.minScrollExtent - position.pixels, 0.0);
    final double overscrollPastEnd =
        math.max(position.pixels - position.maxScrollExtent, 0.0);
    final double overscrollPast =
        math.max(overscrollPastStart, overscrollPastEnd);
    final bool easing = (overscrollPastStart > 0.0 && offset < 0.0) ||
        (overscrollPastEnd > 0.0 && offset > 0.0);

    final double friction = easing
        // Apply less resistance when easing the overscroll vs tensioning.
        ? frictionFactor(
            (overscrollPast - offset.abs()) / position.viewportDimension)
        : frictionFactor(overscrollPast / position.viewportDimension);
    final double direction = offset.sign;
    return direction * _applyFriction(overscrollPast, offset.abs(), friction);
  }

  static double _applyFriction(
      double extentOutside, double absDelta, double gamma) {
    assert(absDelta > 0);
    double total = 0.0;
    if (extentOutside > 0) {
      final double deltaToLimit = extentOutside / gamma;
      if (absDelta < deltaToLimit) return absDelta * gamma;
      total += extentOutside;
      absDelta -= deltaToLimit;
    }
    return total + absDelta;
  }

  @override
  double applyBoundaryConditions(ScrollMetrics position, double value) {
    if (position.maxScrollExtent <= position.pixels && position.pixels < value)
      return value - position.pixels;
    if (position.pixels < position.maxScrollExtent &&
        position.maxScrollExtent < value)
      return value - position.maxScrollExtent;
    return 0.0;
  }

  @override
  Simulation createBallisticSimulation(
      ScrollMetrics position, double velocity) {
    final Tolerance tolerance = this.tolerance;
    if (velocity.abs() >= tolerance.velocity || position.outOfRange) {
      return new BouncingScrollSimulation(
        spring: spring,
        position: position.pixels,
        velocity: velocity * 0.91,
        // TODO(abarth): We should move this constant closer to the drag end.
        leadingExtent: position.minScrollExtent,
        trailingExtent: position.maxScrollExtent,
        tolerance: tolerance,
      );
    }
    return null;
  }

  @override
  double get minFlingVelocity => 2.5 * 2.0;

  @override
  double carriedMomentum(double existingVelocity) {
    return existingVelocity.sign *
        math.min(0.000816 * math.pow(existingVelocity.abs(), 1.967).toDouble(),
            40000.0);
  }

  @override
  double get dragStartDistanceMotionThreshold => 3.5;
}


example/lib/default_loading_layout.dart
import 'package:flutter/material.dart';
import 'package:pull_to_refresh/refresh_header_widget.dart';

class DefaultRefreshHeaderWidget<_DefaultIndicatorState>
    extends RefreshHeaderWidget {
  DefaultRefreshHeaderWidget({Key key}) : super(key: key);

  @override
  _DefaultLoadingState createState() {
    return new _DefaultLoadingState();
  }
}

const int _reset = 0;
const int _releaseToRefresh = 1;
const int _pullToRefresh = 2;
const int _refreshing = 3;

class _DefaultLoadingState extends RefreshState {
  int mode = 0;

  @override
  void reset() {
    setState(() {
      mode = _reset;
    });
  }

  @override
  void onMoving(double offset, double totalHeight, double maxDragHeight) {
    if (offset > maxDragHeight) {
      if (mode != _releaseToRefresh) {
        setState(() {
          mode = _releaseToRefresh;
        });
      }
    } else {
      if (mode != _pullToRefresh) {
        setState(() {
          mode = _pullToRefresh;
        });
      }
    }
  }

  @override
  void refreshing() {
    setState(() {
      mode = _refreshing;
    });
  }

  @override
  Widget build(BuildContext context) {
    Widget textWidget = _buildText();
    Widget iconWidget = _buildIcon();

    List<Widget> children = <Widget>[
      iconWidget,
      new Container(width: 15.0, height: 15.0),
      textWidget
    ];

    return new Container(
      alignment: Alignment.center,
      height: 60.0,
      child: new Center(
        child: new Row(
          textDirection: TextDirection.ltr,
          mainAxisAlignment: MainAxisAlignment.center,
          children: children,
        ),
      ),
    );
  }

  Widget _buildText() {
    switch (mode) {
      case _releaseToRefresh:
        return new Text("Refresh when release");
      case _refreshing:
        return new Text("Refreshing...");
      default:
        return new Text("Pull down to refresh");
    }
  }

  Widget _buildIcon() {
    switch (mode) {
      case _releaseToRefresh:
        return Icon(Icons.arrow_upward, color: Colors.grey);
      case _refreshing:
        return new CircularProgressIndicator(strokeWidth: 2.0);
      default:
        return Icon(Icons.arrow_downward, color: Colors.grey);
    }
  }
}


example/lib/main.dart
import 'dart:async';

import 'package:example/default_loading_layout.dart';
import 'package:flutter/material.dart';
import 'package:pull_to_refresh/pull_to_refresh_widget.dart';
import 'package:flutter/widgets.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool hasMore = true;
  int index = 0;

  List<int> items = List.generate(30, (i) => i);

  Future<Null> _handleRefresh() async {
    await Future.delayed(Duration(seconds: 50), () {
      setState(() {
        items.clear();
        items = List.generate(40, (i) => i);
      });
    });
  }

  Future<Null> _onLoadMore() async {
    await Future.delayed(Duration(seconds: 50), () {
      setState(() {
        if (index == 2) {
          hasMore = false;
        }
        int length = items.length;
        index++;
        items.addAll(List.generate(16, (i) => i + length));
      });
    });
  }

  Widget buildContentWidget(BuildContext context) {
    return PullToRefreshWidget(
        isRefreshEnable: true,
        headerBuilder: _buildHeaderWidget,
        onRefresh: _handleRefresh,
        isLoadMoreEnable: hasMore,
        onLoadMore: _onLoadMore,
        footerBuilder: _buildFootWidget,
        child: ListView.builder(
            itemCount: items.length,
            itemBuilder: (context, index) {
              return new ListTile(
                title: Text("Index$index"),
              );
            }));
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
        appBar: new AppBar(
          // Here we take the value from the MyHomePage object that was created by
          // the App.build method, and use it to set our appbar title.
          title: new Text(widget.title),
        ),
        body: new Container(child: buildContentWidget(context)));
  }

  DefaultRefreshHeaderWidget _buildHeaderWidget(BuildContext context) {
    return DefaultRefreshHeaderWidget();
  }

  Widget _buildFootWidget(BuildContext context) {
    return new Container(
      alignment: Alignment.center,
      height: 90.0,
      child: new Center(
        child: new Row(
            textDirection: TextDirection.ltr,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              new Text("正在加载中。。。"),
              new CircularProgressIndicator(strokeWidth: 2.0)
            ]),
      ),
    );
  }
}
